server/routerlicious/packages/services/src/throttlerHelper.ts

```diff
@@ -0,0 +1,117 @@
+            // update stored throttling metric with new retry duration
+            await this.throttleStorageManager.setThrottlingMetric(id, throttlingMetric);
+            return this.getThrottlerResponseFromThrottlingMetrics(throttlingMetric);
+        }
+
+        // replenish "tokens" if possible
+        const amountToReplenish = this.getTokenReplenishAmount(throttlingMetric, now);
+        if (amountToReplenish > 0) {
+            throttlingMetric.count += amountToReplenish;
+            throttlingMetric.lastCoolDownAt = now;
+        }
+
+        // adjust "tokens" based on given count
+        throttlingMetric.count -= count;
+
+        // throttle if "token bucket" is empty
+        const newTimeUntilNotThrottled = this.getRetryAfterInMs(throttlingMetric, now);
+        if (newTimeUntilNotThrottled > 0) {
+            throttlingMetric.throttleStatus = true;
+            throttlingMetric.throttleReason =
+                `Throttling count exceeded by ${Math.abs(throttlingMetric.count)} at ${new Date(now).toISOString()}`;
+            throttlingMetric.retryAfterInMs = newTimeUntilNotThrottled;
+        } else {
+            throttlingMetric.throttleStatus = false;
+            throttlingMetric.throttleReason = "";
+            throttlingMetric.retryAfterInMs = 0;
+        }
+
+        // update stored throttling metric
+        await this.throttleStorageManager.setThrottlingMetric(id, throttlingMetric);
+
+        return this.getThrottlerResponseFromThrottlingMetrics(throttlingMetric);
+    }
+
+    public async getThrottleStatus(id: string): Promise<IThrottlerResponse | undefined> {
+        const throttlingMetric = await this.throttleStorageManager.getThrottlingMetric(id);
+        if (!throttlingMetric) {
+            return undefined;
+        }
+        return this.getThrottlerResponseFromThrottlingMetrics(throttlingMetric);
+    }
+
+    private getThrottlerResponseFromThrottlingMetrics(throttlingMetric: IThrottlingMetrics): IThrottlerResponse {
+        return {
+            throttleStatus: throttlingMetric.throttleStatus,
+            throttleReason: throttlingMetric.throttleReason,
+            retryAfterInMs: throttlingMetric.retryAfterInMs,
+        };
+    }
+
+    private getTokenReplenishAmount(throttlingMetric: IThrottlingMetrics, now: number): number {
+        const timeSinceLastCooldownInMs = now - throttlingMetric.lastCoolDownAt;
+        // replenish "tokens" at most once per minCooldownInterval
+        if (timeSinceLastCooldownInMs > this.minCooldownIntervalInMs) {
+            return Math.floor(timeSinceLastCooldownInMs * this.rateInOperationsPerMs);
+        }
+        return 0;
+    }
+
+    private getRetryAfterInMs(throttlingMetric: IThrottlingMetrics, now: number): number {
+        const tokenDebt = 0 - throttlingMetric.count;
+        if (tokenDebt <= 0) {
+            // no wait time is necessary before next retry because token bucket has enough tokens
```

Comment:
isn't this condition always true?

Reply:
If the operation count exceeds the allowed rate, then count will be reduced below 0 on line 58. In other words, if the ops happen faster than the token cooldown, count will drop below zero, thus triggering the throttle.